### PR TITLE
chore(ci): add bench comparing remap and other implementations

### LINKED ISSUE #5945: Add remap comparison to benches
We have benches comparing wasm vs. lua vs. native transforms: https://github.com/timberio/vector/blob/0dd3fe2c1c1c84e5ad6925f1e356dace1a6795a4/benches/wasm/mod.rs#L60-L115

It'd be useful to add remap for comparison to these benches.

### PR DESCRIPTION
**Pull request by jszwedko**
Opening as a draft as I still have a couple of TODOs, but this PR:

* Adds remap to the wasm bench that compares native vs. wasm vs. lua (`add_fields`) to get a rough idea of the raw overhead of using remap
* Uses a more complete example script in a topology bench to get a more "real world comparison"

TODOs:
* [x] (Maybe) update bench to generate random lines rather than static. Doing this does introduce some inherent noise into the bench though.
* [x] Modify example to one that can be completely implemented with native transforms for a direct comparison (current example generates UUIDs which we don't have a native transform for)
* [x] Add debug assertions to ensure that the topologies actually transform the events in the expected ways
* [x] Fix WASM test assertion for the added module. This could be done by removing uuid generation which is the dynamic part which stops us from doing a static input / output comparison.
* [x] Refactor bench to share scaffolding code to make it easier to add additional benches

At the risk of discovering there is something wrong with the bench when I add debug assertions, this is what it currently looks like for the VRL example script (translated into lua and wasm):

```
   . = parse_syslog(.message)
    .severity = "info"
    .id = uuid_v4()
    .timestamp = to_int(.timestamp)
```

Using a tcp socket for input and output. I initially thought I could test the pipeline of transforms without that, but couldn't find a good avenue that still used the topology code. I'm open to ideas here.

2021-01-14-183929_1919x510_scrot

Closes #5945

### CODE REVIEW

**Comment by jszwedko** [comment]
One thing I realized these benches don't really encapsulate is the advantage we'll see of `remap` being a `FunctionTransform` that is able to be pushed up and parallelized in a way that the `wasm` and `lua` cannot currently. It'd be good to add a benchmark for that when that work happens.

**Comment by jszwedko** [comment]
Comparison for `add_fields` which shows that remap starts around 3.5x faster, but does not see increases due to input size (number of key/value pairs in the event being sent in) that the wasm and lua runtimes see:

2021-01-14-195538_1819x923_scrot

File: benches/language.rs [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    local template ='xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'
+    return string.gsub(template, '[xy]', function (c)
+        local v = (c == 'x') and random(0, 0xf) or random(8, 0xb)
+        return string.format('%x', v)
+    end)
+end
+
+function process(event, emit)
+  event.log.id = uuid()
+  event.log.timestamp = os.time(event.log.timestamp)
```
Comment (JeanMertz) [inline benches/language.rs:99]: Doesn’t the native part already take care of this?

File: benches/language.rs [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+                    rt.block_on(async move {
+                        send_lines(in_addr, lines).await.unwrap();
+
+                        topology.stop().compat().await.unwrap();
+
+                        let output_lines = output_lines.await;
+
+                        #[cfg(debug_assertions)]
+                        {
+                            // TODO assert lines are tranformed
```
Comment (JeanMertz) [inline benches/language.rs:237]: Yeah. We should definitely make sure of this. Just in case one of the examples isn’t wired up correctly.

**Review by JeanMertz — COMMENTED** [review]
Nice! This is looking really good 👌

The fact that Wasm is so much slower than Lua makes me curious though. I know LuaJit is fast, but still...

File: benches/language.rs (hunk 1 above)
Comment (jszwedko) [reply benches/language.rs:99]: This is actually converting it to a unix integer timestamp.

**Comment by jszwedko** [comment]
@JeanMertz thanks for taking a look!

> Nice! This is looking really good ok_hand
> 
> The fact that Wasm is so much slower than Lua makes me curious though. I know LuaJit is fast, but still...

Yeah, I think this is largely because the `add_fields` logic is very simple/small so most of the time is spent serializing the event and passing it back and forth. This appears to be more expensive in our wasm implementation (there are probably areas for improvement) than our Lua implementation.

**Comment by JeanMertz** [comment]
> Yeah, I think this is largely because the `add_fields` logic is very simple/small so most of the time is spent serializing the event and passing it back and forth.

That's true, but the "VRL announcement" one is more involved, and it's still twice as slow.

But yeah, I remember @Hoverbear talking about how it currently serializes to JSON to pass the data around, which obviously incurs a high overhead when lots of data is pushed back and forth.

**Comment by Hoverbear** [comment]
Correct! We use JSON right now as we weren't sure what our target languages were for WASM, @lukesteensen and I discussed using CBOR or protobuf, which should be much faster as it can skip a lot of work, but that was left for future work.

**Comment by binarylogic** [comment]
@jszwedko it would be nice to see comparisons for:

1. Baseline (ex: adding fields)
2. Parsing JSON
3. Parsing Syslog
4. Redaction (something simple like credit cards or emails)
5. Combined (parsing, reshaping, redacting)

That would be a strong showing for comparison across all runtimes. Is this doable?

**Comment by lucperkins** [comment]
@binarylogic I can add those use cases once this gets through.

**Comment by jszwedko** [comment]
:+1: I think we can certainly expand these to cover those cases (though not all will be supportable with just native transforms). I'd like to just merge this with an initial example and open a separate issue for expanding it to cover more use-cases. It sounds like @lucperkins would be interested in taking that on; I'm happy to assist.

@MOZGIII this bench is an example of something where I'd like to drive a topology (set of transforms) without involving sockets which seems related to our conversation earlier. I'm curious if you have any thoughts. I couldn't see an easy way.

**Comment by MOZGIII** [comment]
I'd suggest that we provide a special, bench-oriented source and a bench-oriented sink for such use cases. That is if we want to test the transform as part of topology specifically, and not the raw transform speed itself.

To test the transform in isolation, we don't really need to wire it up through the topology - it's enough to explicitly construct it, and give it explicitly build in/out channels (rather than the ones assembled as part of the topology).
It can easily turn out that some of our transforms are way quicker than our topology allows them to operate, and thus the topology is the bottleneck. Assuming it won't always be the case (say we improve how the topology performs), we should aim to isolate the actual transform-specific code from the rest of the concerns as much as possible.

**Comment by jszwedko** [comment]
I longer than I should have going down the rabbit hole of using a mock source/sink implementation to avoid sockets, but this is not as straightforward as I expected due to some assumptions that the topology code makes about the way that source components work. Sinks should be relatively straightforward on the other hand.

After talking with @MOZGIII I decided to punt on that for this PR so we could focus on it later as part refactoring the topology layer to be more benchable.

I've started to move forward with modifying the example to be one that can completely be implemented as native transforms for comparison. I'll aim to wrap that and the other TODOs here tomorrow.

**Comment by jszwedko** [comment]
I think this is ready for review now. I:

* refactored a bit to, hopefully, make it simpler to add additional language comparison benchmarks
* added debug assertions to ensure events are transformed
* simplified the example to just be syslog message parsing so that I could implement it with `native` transforms too for comparison

File: tests/data/wasm/assert_config/fixtures/a/expected.json [hunk 3]
```diff
@@ -11,6 +11,6 @@
     "takes_number": 123,
     "takes_string": "test"
   },
-  "path": "tests/data/wasm/protobuf/target/wasm32-wasi/release/protobuf.wasm",
+  "path": "tests/data/wasm/assert_config/target/wasm32-wasi/release/assert_config.wasm",
```
Comment (jszwedko) [inline tests/data/wasm/assert_config/fixtures/a/expected.json:14]: Noticed this was broken when running the tests.

File: benches/language.rs [hunk 4]
```diff
@@ -0,0 +1,193 @@
+use criterion::{criterion_group, criterion_main, BatchSize, Criterion, SamplingMode, Throughput};
```
Comment (MOZGIII) [inline benches/language.rs:1]: Let's rename this file to `benches/languages.rs` for consistency with `make bench-languages`. It's a small nit, but I got bit by this, so I'm proposing it.

**Comment by JeanMertz** [comment]
Are there any updates numbers to share @jszwedko ?

**Comment by jszwedko** [comment]
@JeanMertz I ended up simplifying the example to:

```
. = parse_syslog!(.message)
```

To be able to implement the transformation with native transforms in addition to lua and wasm. This is what it looks like:

2021-01-22-143909_1472x430_scrot

Native transforms win, but notably it is using `regex_parser` which might be faster than `syslog_loose`. I think adding more complex examples, even if they don't have native transform analogues, will be useful. I'll leave that as a future issue though: https://github.com/timberio/vector/issues/6198

**Comment by jszwedko** [comment]
@MOZGIII feel free to merge this if you agree with it.

**Comment by lucperkins** [comment]
It looks like there was a feature naming issue, hence the CI failure. I've pushed a fix for that.

**Pull request merged**